### PR TITLE
7903596: JMH: Explicitly enable annotation processors

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review of this change which proposes to address the issue noted in https://bugs.openjdk.org/browse/CODETOOLS-7903596?

As noted in that issue, when a benchmark generated using Java 22 is run, at runtime it fails with:

```console
java -jar target/benchmarks.jar
Exception in thread "main" java.lang.RuntimeException: ERROR: Unable to find the resource: /META-INF/BenchmarkList
	at org.openjdk.jmh.runner.AbstractResourceReader.getReaders(AbstractResourceReader.java:98)
	at org.openjdk.jmh.runner.BenchmarkList.find(BenchmarkList.java:124)
	at org.openjdk.jmh.runner.Runner.internalRun(Runner.java:252)
	at org.openjdk.jmh.runner.Runner.run(Runner.java:208)
	at org.openjdk.jmh.Main.main(Main.java:71)

```
This is due to annotation processors being disabled by default during compilation starting Java 22. The commit in this PR, explicitly enables the `org.openjdk.jmh.generators.BenchmarkProcessor` annotation processor to allow for it to generate the necessary benchmark resources.

I've tested this change locally, with Java 8, Java 21 as well as Java 22 and the newly generated benchmark is now functional on all these versions.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [CODETOOLS-7903596](https://bugs.openjdk.org/browse/CODETOOLS-7903596): JMH: Explicitly enable annotation processors (**Bug** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - Committer)

### Contributors
 * Aleksey Shipilev `<shade@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmh.git pull/125/head:pull/125` \
`$ git checkout pull/125`

Update a local copy of the PR: \
`$ git checkout pull/125` \
`$ git pull https://git.openjdk.org/jmh.git pull/125/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 125`

View PR using the GUI difftool: \
`$ git pr show -t 125`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmh/pull/125.diff">https://git.openjdk.org/jmh/pull/125.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jmh/pull/125#issuecomment-1833201692)
</details>

### CODE REVIEW

**Comment by shipilev** [comment]
There are tests in JMH that test archetypes work, please run them. The easiest way is do this is enabling JMH GHA: 
https://github.com/jaikiran/jmh/actions -- and triggering the run on your branch.

**Comment by jaikiran** [comment]
Hello Aleksey, thank you for that info. I've enabled the actions and the job is in progress. There are some Windows failures, but those don't appear related:

```
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 177.957 sec <<< FAILURE! - in org.openjdk.jmh.it.profilers.WinPerfAsmProfilerTest
test(org.openjdk.jmh.it.profilers.WinPerfAsmProfilerTest)  Time elapsed: 177.832 sec  <<< ERROR!
java.lang.IllegalStateException: Profile does not contain the required frame: PrintAssembly processed: 7 total address lines.
```

There's a macos run going on in that run for a while now, I'll check it out tomorrow on how it goes.

**Comment by jaikiran** [comment]
I had a look at the Windows failures reported in GitHub actions job - they all appear to be:

```
test(org.openjdk.jmh.it.profilers.WinPerfAsmProfilerTest)  Time elapsed: 185.743 sec  <<< ERROR!
java.lang.IllegalStateException: Profile does not contain the required frame: PrintAssembly processed: 7 total address lines.
```
I'm unfamiliar with those tests, but the error doesn't suggest it's due to the changes in this PR. Would you happen to know, Aleksey?

**Comment by jaikiran** [comment]
Please keep open, hoping for a review.

**Comment by jaikiran** [comment]
Re-opening this PR since the issue is still applicable.

Without the changes in this PR, generating a project using the archetype:

**Comment by jaikiran** [comment]
Re-opening this PR since the issue is still applicable.

Without the changes in this PR, generating a project using the archetype:

```
mvn archetype:generate \       
  -DinteractiveMode=false \
  -DarchetypeGroupId=org.openjdk.jmh \
  -DarchetypeArtifactId=jmh-java-benchmark-archetype -DarchetypeVersion=1.38-SNAPSHOT\
  -DgroupId=org.sample \
  -DartifactId=test \
  -Dversion=1.0
```

Then running the project:

```
cd test
mvn clean install
$JDK23/bin/java -jar target/benchmarks.jar
```
results in:

```
Exception in thread "main" java.lang.RuntimeException: ERROR: Unable to find the resource: /META-INF/BenchmarkList
	at org.openjdk.jmh.runner.AbstractResourceReader.getReaders(AbstractResourceReader.java:98)
	at org.openjdk.jmh.runner.BenchmarkList.find(BenchmarkList.java:124)
	at org.openjdk.jmh.runner.Runner.internalRun(Runner.java:252)
	at org.openjdk.jmh.runner.Runner.run(Runner.java:208)
	at org.openjdk.jmh.Main.main(Main.java:71)
```

With the changes in this PR, the newly generated projects from the archetype can successfully launch the benchmarks even against Java 23.

**Comment by sigpwned** [comment]
I just got bit by this issue and lost a couple of hours. Having this archetype up-to-date would have helped. Instead, my developer experience looked like this:

1. Try to add JMH benchmarks to a new project
2. Consistently get above error about /META-INF/BenchmarkList
3. Try to run existing JMH benchmarks on existing projects
4. Those fail with /META-INF/BenchmarkList
5. Generate a fresh project from the archetype
6. Try to run JMH benchmarks against the archetype
7. Those fail with /META-INF/BenchmarkList
8. Bang head against wall
9. Check JMH PRs and find this issue
10. Retry builds with Java 17, and everything works

First, I could have saved some time if this archetype had been up-to-date, since it would have given me the thread of the relevant difference in the POM to pull on. Second, that I'm only discovering this through a PR I had to track down semi-manually means that there are probably a lot of folks having issues who can't find the cause.

In my opinion, for whatever that's worth, I think merging this PR would be very helpful, since it would at least put this PR in the main branch to help people find it, and I also think adding a "gotcha" to the README would help, too.

**Comment by galderz** [comment]
Just bumped into this.

> I also think adding a "gotcha" to the README would help, too.

+1

**Comment by jaikiran** [comment]
Any reviewers who can review this one, please?

**Review by shipilev — COMMENTED** [review]
This looks okay, but shouldn't the fix also include adding `-processor` all around the JMH modules, like `jmh-samples`, `jmh-core-it` and such? Should probably be enough to add it to `jmh-parent` here: https://github.com/openjdk/jmh/blob/7c6da9526c6544acbe9f19f2cf075793f240b120/pom.xml#L175-L179

Please rebase from master and test full JMH build with JDK 23.

**Comment by shipilev** [comment]
All right, I think this is what we minimally need to be able to build and test on JDK 23: [jmh-jdk23-1.patch](https://github.com/user-attachments/files/18028639/jmh-jdk23-1.patch).

@sigpwned -- `jmh-ct-tests` are weird in the sense they use their own in-memory compiler, so adding things to POM would not work :) It is also why they are compiled with `-proc:none`. See the patch above where the in-memory compiler options are added. Good call on using `<annotationProcessors>` in POMs, though, it is a bit cleaner.

@jaikiran, please mix it in this PR, or shout if you don't have time to complete this PR, and I'll take over.

**Comment by jaikiran** [comment]
Hello Aleksey,

> @jaikiran, please mix it in this PR, or shout if you don't have time to complete this PR, and I'll take over.

I have been a bit busy with something else and haven't been able to get to this. I'll apply your patch into this PR today. Thank you for the help.

File: jmh-archetypes/jmh-java-benchmark-archetype/src/main/resources/archetype-resources/pom.xml [hunk 1]
```diff
@@ -88,6 +88,10 @@ THE POSSIBILITY OF SUCH DAMAGE.
                     <compilerVersion>\${javac.target}</compilerVersion>
                     <source>\${javac.target}</source>
                     <target>\${javac.target}</target>
+                    <compilerArgs>
+                        <arg>-processor</arg>
+                        <arg>org.openjdk.jmh.generators.BenchmarkProcessor</arg>
+                    </compilerArgs>
```
Comment (shipilev) [inline jmh-archetypes/jmh-java-benchmark-archetype/src/main/resources/archetype-resources/pom.xml:94]: Forgot to mention this: this should probably be using `<annotationProcessors>` as well, like the other hunks.

**Review by shipilev — COMMENTED** [review]
Thanks! Please fix the following nit and merge from master to get cleaner macos runs.

**Review by shipilev — APPROVED** [review]
Looks good. Feel free to integrate after testing turns green.

**Comment by shipilev** [comment]
I think MacOS test failure is ignorable.

**Comment by jaikiran** [comment]
Thank you Aleksey and others for the help on this one. I haven't been able to run a complete test of these changes locally (due to unrelated environmental issues), but given what Aleksey noted about the GH actions job, I'll go ahead with the integration.

**Pull request closed without merging**